### ISSUE TITLE
docker swarm init fails on IPv6 only host

### ISSUE THREAD

**Issue by calvincheng8** [docker/docker#23995]
**Output of `docker version`:**

```
Client:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd-unsupported
 Built:        Fri Jun 17 21:21:56 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd-unsupported
 Built:        Fri Jun 17 21:21:56 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 1
Server Version: 1.12.0-rc2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 33
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host overlay bridge null
Swarm: inactive
Runtimes: default
Default Runtime: default
Security Options: apparmor seccomp
Kernel Version: 4.4.0-21-generic
Operating System: Ubuntu 16.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 15.67 GiB
Name: bn831x1-ossv-3a-lab
ID: B272:E7BR:NRDW:NJ5R:TTAX:U27E:NML7:AT7G:AV4Z:OZG3:RTXH:MQGL
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
My host is a KVM guest running Ubuntu 16.04 LTS, with only IPv6 networking enabled. No IPv4 is configured.

**Steps to reproduce the issue:**
1. Start dockerd with "-H fd:// --ipv6=true" options
2. Run "docker swarm init"

**Describe the results you received:**
Receives the follower error due to missing IPv4 network.
```
# docker swarm init
Error response from daemon: could not determine local IP address: dial udp 8.8.8.8:53: connect: network is unreachable
```

**Describe the results you expected:**
No error with swarm initialized

**Additional information you deem important (e.g. issue happens only occasionally):**
Reproducible on different hosts each time.

**Comment by cpuguy83** [comment]
Looking at this, it looks like it should just work if you specify your own listen addr.
`docker swarm init --listen-addr <ip>:<port>`

Otherwise it tries to infer a listen address based on the default route, which it accomplishes by dialing out.
I'll see if there's a better way to handle this.

**Comment by calvincheng8** [comment]
Just tried
```
$ docker swarm init --listen-addr [2003:1234::1:1]:2377
invalid argument  [2003:1234::1:1]:2377" for --listen-addr: Invalid url, too many colons
See 'docker swarm init --help'.
```

Switching to hostname with IPv6 defined in /etc/hosts:
```
2003:1234::1:1 myhost
```

And it worked!

```
$ docker swarm init --listen-addr myhost:2377
Swarm initialized: current node (f2j9aknwtum0e936pj2u7nrzz) is now a manager.
```

Please fix the parsing of IPv6 address as well as looking up the default route when IPv4 is disabled.

**Comment by tonistiigi** [comment]
Parsing should be fixed after #23832

**Comment by icecrime** [comment]
Ping @aaronlehmann: this relates to docker/swarmkit#631.

**Comment by cpuguy83** [comment]
I have a patch for getting the default route, just mucking around with godep at the moment.

**Comment by aaronlehmann** [comment]
@cpuguy83: Curious to see your patch. Note that the current dialing out uses UDP and never sends any data, so all it is really doing is allocatibg a socket and using that socket to infer the IP address associated with the default route. We could add a fallback that uses IPv6 for machines without IPv4.

The thing I like about this dial-based detection is that it should be completely portable. I don't know of a portable way to do this by inspecting the routing table.

**Comment by cpuguy83** [comment]
@aaronlehmann Yes, my method is not portable, but I have it guarded by a platform build flag. I'll post it as soon as I get the hang of godeps here, we'll see if it's worth it.

**Comment by thaJeztah** [comment]
@cpuguy83 got the hang of godeps already? :wink:

**Comment by aaronlehmann** [comment]
This will be fixed by #24237.
